src/ngraph/runtime/cpu/pass/cpu_mat_fusion.cpp

```diff
@@ -199,3 +201,104 @@ bool runtime::cpu::pass::CPURnnMatFusion::run_on_function(std::shared_ptr<Functi
+        auto skip = std::make_shared<pattern::op::Skip>(slice, reshape_pred);
+        reshape[i] = std::make_shared<op::Reshape>(skip, AxisVector{0, 1, 2}, Shape{2, 2});
+    }
+    auto dot = std::make_shared<op::Dot>(reshape[0], reshape[1]);
+    auto final_reshape = std::make_shared<op::Reshape>(dot, AxisVector{0, 1}, Shape{1, 2, 2});
+
+    auto matcher = std::make_shared<pattern::Matcher>(final_reshape);
+    std::shared_ptr<Node> fuse_input[num_op_branches];
+    bool transpose[num_op_branches] = {false, false};
+    const int num_expected_reshape_with_trans = 3;
+
+    // check each input arg matches the pattern
+    for (auto arg : n->get_arguments())
+    {
+        if (matcher->match(arg))
+        {
+            auto pattern_map = matcher->get_pattern_map();
+            int reshape_count[num_op_branches] = {0, 0};
+            // we found a match, determine whether we have to transpose for each input by
+            // counting the number of reshapes in each branch, if transpose is applied, there
+            // should be 3 reshapes.
+            for (int i = 0; i < num_op_branches; ++i)
+            {
+                auto iter = matcher->get_match_root();
+                auto& input_node = pattern_map[input[i]];
+                do
+                {
+                    if (std::dynamic_pointer_cast<op::Reshape>(iter) != nullptr)
+                    {
+                        ++reshape_count[i];
+                        if (reshape_count[i] == num_expected_reshape_with_trans)
+                        {
+                            transpose[i] = true;
+                            break;
+                        }
+                    }
+                    // branch to either input 0 or 1 depending on which one we are traversing
+                    iter =
+                        iter->get_input_size() > 1 ? iter->get_argument(i) : iter->get_argument(0);
+                } while (iter != input_node);
+            }
+            // keep track of the input data, make sure they all match
+            for (int i = 0; i < num_op_branches; ++i)
+            {
+                auto& input_node = pattern_map[input[i]];
+                if (fuse_input[i] == nullptr)
+                {
+                    fuse_input[i] = input_node;
+                }
+                // found different input nodes between different args, can't fuse.
+                else if (fuse_input[i] != input_node)
+                {
+                    return {nullptr};
+                }
+            }
+        }
+    }
+    if (fuse_input[0] && fuse_input[1])
+    {
+        return std::make_shared<op::BatchDot>(
+            fuse_input[0], fuse_input[1], transpose[0], transpose[1]);
+    }
+    return {nullptr};
+}
+
+bool runtime::cpu::pass::CPUBatchDotFusion::run_on_function(std::shared_ptr<Function> func)
+{
+    bool modified = false;
+
+    for (auto n : func->get_ordered_ops())
+    {
+        const Node& node = *n;
```

Comment:
if we wud just do `std::dynamic_pointer_cast<op::Concat>` we could avoid all this `*n` and `TI` class comparisons 😃

Reply:
Copy&paste our conversation: Dynamic cast actually works for derived class too. So there's potentially ambiguity in this check and we want `concat` exactly, not potentially a derived class of `concat`